### PR TITLE
FCIT

### PR DESCRIPTION
**Pull request by MatthewZhao26**

### CODE REVIEW

**Comment by MatthewZhao26** [comment]
@sampan501 finished documentation, ready for comments

File: hyppo/__init__.py [hunk 1]
```diff
@@ -5,5 +5,6 @@
 import hyppo.kgof
 import hyppo.tools
 import hyppo.d_variate
+import hyppo.conditional_independence
```
Comment (sampan501) [inline hyppo/__init__.py]: rename it conditional

File: hyppo/conditional_independence/FCIT.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+       Dictionary of parameters to cross-validate over when training regressor.
+    num_perm: int
+        Number of data permutations to estimate the p-value from marginal stats.
+    prop_test: float
+        Proportion of data to evaluate test stat on.
+    discrete: tuple of string
+        Whether :math: `X` or :math: `Y` are discrete
+    Notes
+    -----
+    The motivation for the test rests on the assumption that if :math: `X \not\!\perp\!\!\!\perp Y \mid Z`,
```
Comment (sampan501) [inline hyppo/conditional_independence/FCIT.py]: This is not rendering correctly in the documentation. See: https://deploy-preview-315--hyppo.netlify.app/api/generated/hyppo.conditional_independence.fcit#hyppo.conditional_independence.FCIT

File: hyppo/conditional_independence/test_delete_later.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
+import numpy as np
```
Comment (sampan501) [inline hyppo/conditional_independence/test_delete_later.py]: reformat this into classes like other tests in hyppo

File: hyppo/conditional_independence/tests/test_FCIT.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import numpy as np
```
Comment (sampan501) [inline hyppo/conditional_independence/tests/test_FCIT.py]: what's the difference between the previous test file and this one?

File: hyppo/conditional_independence/test_delete_later.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+z = np.random.randn(500, 10)
+
+
+# print(FCIT(model=model, cv_grid=cv_grid).test(x, y, z))
+
+# np.random.seed(123456789)
+# x, y = rot_ksamp("linear", 1000, 1, k=2)
+
+# np.random.seed(123456789)
+# print(FCIT(model=model, cv_grid=cv_grid).test(x,y))
```
Comment (sampan501) [inline hyppo/conditional_independence/test_delete_later.py]: the __init__.py file should not be empty, see the independence __init__.py

File: hyppo/tools/tests/test_common.py [hunk 6]
```diff
@@ -185,7 +185,7 @@ def test_diskern(self):
                 l1, (n - 1, n + 1), (l1.itemsize * (n + 1), l1.itemsize)
```
Comment (sampan501) [inline hyppo/tools/tests/test_common.py]: why are you modifying this file?

File: hyppo/tools/ksample_sim.py [hunk 7]
```diff
@@ -36,7 +36,7 @@ def _2samp_rotate(sim, x, y, p, degree=90, pow_type="samp"):
         ]:
```
Comment (sampan501) [inline hyppo/tools/ksample_sim.py]: why are you modifying this file?

File: hyppo/tools/indep_sim.py [hunk 8]
```diff
@@ -195,9 +195,9 @@ def cubic(n, p, noise=False, low=-1, high=1, cubs=[-12, 48, 128], scale=1 / 3):
```
Comment (sampan501) [inline hyppo/tools/indep_sim.py]: why are you modifying this file?

File: hyppo/tools/common.py [hunk 9]
```diff
@@ -178,7 +178,7 @@ def compute_kern(x, y, metric="gaussian", workers=1, **kwargs):
             )
```
Comment (sampan501) [inline hyppo/tools/common.py]: why are you modifying this file?

File: hyppo/time_series/tests/test_mgcx.py [hunk 10]
```diff
@@ -44,7 +44,7 @@ def test_optimal_scale_linear(self):
     def test_optimal_scale_nonlinear(self):
```
Comment (sampan501) [inline hyppo/time_series/tests/test_mgcx.py]: why are you modifying this file?

File: hyppo/ksample/mmd.py [hunk 11]
```diff
@@ -112,7 +112,7 @@ def statistic(self, x, y):
                 x,
```
Comment (sampan501) [inline hyppo/ksample/mmd.py]: why are you modifying this file?

File: hyppo/ksample/mean_embedding.py [hunk 12]
```diff
@@ -166,7 +166,7 @@ def mean_embed_distance(difference, num_randfeatures):
     mu = np.mean(difference, 0)
```
Comment (sampan501) [inline hyppo/ksample/mean_embedding.py]: why are you modifying this file?

File: hyppo/ksample/energy.py [hunk 13]
```diff
@@ -115,7 +115,7 @@ def statistic(self, x, y):
```
Comment (sampan501) [inline hyppo/ksample/energy.py]: why are you modifying this file?

File: hyppo/kgof/kernel.py [hunk 14]
```diff
@@ -162,8 +162,8 @@ def eval(self, X, Y):
         # (n1, d1) = X.shape
```
Comment (sampan501) [inline hyppo/kgof/kernel.py]: why are you modifying this file?

File: hyppo/kgof/fssd.py [hunk 15]
```diff
@@ -381,7 +381,7 @@ def ustat_h1_mean_variance(fea_tensor, return_variance=True, use_unbiased=True):
     Tau = np.reshape(Xi, [n, d * J])
```
Comment (sampan501) [inline hyppo/kgof/fssd.py]: why are you modifying this file?

File: hyppo/kgof/_utils.py [hunk 16]
```diff
@@ -92,7 +92,7 @@ def fit_gaussian_draw(X, J, seed=28, reg=1e-7, eig_pow=1.0):
     evals = np.maximum(0, np.real(evals))
```
Comment (sampan501) [inline hyppo/kgof/_utils.py]: why are you modifying this file?

File: hyppo/independence/dcorr.py [hunk 17]
```diff
@@ -378,7 +378,7 @@ def _fast_1d_dcov(x, y, bias=False):  # pragma: no cover
     ) + (si[-1] - 2 * si)
```
Comment (sampan501) [inline hyppo/independence/dcorr.py]: why are you modifying this file?

File: hyppo/independence/cca.py [hunk 18]
```diff
@@ -73,8 +73,8 @@ def statistic(self, x, y):
```
Comment (sampan501) [inline hyppo/independence/cca.py]: why are you modifying this file?

File: hyppo/d_variate/dhsic.py [hunk 19]
```diff
@@ -101,10 +101,10 @@ def statistic(self, *args):
         term3 = (2 / n) * np.ones((n,))
```
Comment (sampan501) [inline hyppo/d_variate/dhsic.py]: why are you modifying this file?

Comment (sampan501) [reply hyppo/d_variate/dhsic.py]: also make sure your black formatting is up to date

**Review by sampan501 — CHANGES_REQUESTED** [review]

File: hyppo/__init__.py (hunk 1 above)
Comment (MatthewZhao26) [reply hyppo/__init__.py]: @sampan501 renamed

File: hyppo/conditional_independence/test_delete_later.py (hunk 3 above)
Comment (MatthewZhao26) [reply hyppo/conditional_independence/test_delete_later.py]: @sampan501 originally meant this as a scratch file; actual test is in test folder, will delete in next push

File: hyppo/conditional_independence/tests/test_FCIT.py (hunk 4 above)
Comment (MatthewZhao26) [reply hyppo/conditional_independence/tests/test_FCIT.py]: see above

File: hyppo/tools/tests/test_common.py (hunk 6 above)
Comment (MatthewZhao26) [reply hyppo/tools/tests/test_common.py]: @sampan501 I ran black and it automatically made these and all following changes

File: hyppo/conditional_independence/test_delete_later.py (hunk 5 above)
Comment (MatthewZhao26) [reply hyppo/conditional_independence/test_delete_later.py]: @sampan501 my init.py file for conditional is not empty, could this comment have been made on the wrong file (it's associated with the test_delete_later.py file)?

**Comment by MatthewZhao26** [comment]
@sampan501 just addressed all comments

File: hyppo/conditional/FCIT.py [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+    ):
+
+        self.model = model
+        self.cv_grid = cv_grid
+        self.num_perm = num_perm
+        self.prop_test = prop_test
+        self.discrete = discrete
+        ConditionalIndependenceTest.__init__(self)
+
+    def statistic(self, x, y, z=None):
```
Comment (sampan501) [inline hyppo/conditional/FCIT.py:72]: This function should only compute the test statistic. Any p-value computation should be moved to the `.test` method

File: hyppo/conditional/FCIT.py [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+            x, y = y, x
+
+        y = StandardScaler().fit_transform(y)
+
+        stat, pvalue = self.statistic(x, y, z)
+
+        return ConditionalIndependenceTestOutput(stat, pvalue)
+
+
+def cross_val(x, y, z, cv_grid, model, prop_test):
```
Comment (sampan501) [inline hyppo/conditional/FCIT.py]: Since these are private functions, they should start with `_`

**Review by sampan501 — CHANGES_REQUESTED** [review]
Where is the tutorial for this code?

**Comment by sampan501** [comment]
not sure why you don't see the deploy review? I can see it in the checks

**Comment by MatthewZhao26** [comment]
> not sure why you don't see the deploy review? I can see it in the checks

@sampan501 I made documentation but I'm not sure which file to check to make sure it shows up on the site

**Comment by sampan501** [comment]
> @sampan501 I made documentation but I'm not sure which file to check to make sure it shows up on the site

https://github.com/neurodata/hyppo/blob/dev/docs/api/index.rst

**Comment by MatthewZhao26** [comment]
> > @sampan501 I made documentation but I'm not sure which file to check to make sure it shows up on the site
> 
> https://github.com/neurodata/hyppo/blob/dev/docs/api/index.rst

@sampan501  Ok, I already did add the conditional module to this file. Maybe it's not showing up due to the error in the tutorial code I mentioned on Slack? I'll try pushing again

File: docs/api/index.rst [hunk 22]
```diff
@@ -87,6 +87,18 @@ Discriminability
    DiscrimOneSample
    DiscrimTwoSample
 
+.. automodule:: hyppo.conditional
+
+.. currentmodule:: hyppo.conditioal
+
+
+Conditional Independence
+_________________________
```
Comment (sampan501) [inline docs/api/index.rst]: This is not the same style of heading, which is why it is not showing up. Should be hyphenated, i.e. `---`. See [this](https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html#headings) for ordering

**Review by sampan501 — COMMENTED** [review]
mentioned the fix here

File: docs/api/index.rst (hunk 22 above)
Comment (MatthewZhao26) [reply docs/api/index.rst]: got it, changing now

**Comment by MatthewZhao26** [comment]
> mentioned the fix here

and under the tutorials folder, I'd also have to make a new file titled "conditional.py" right" @sampan501

**Comment by MatthewZhao26** [comment]
@sampan501 Addressed all comments and added tutorial

**Comment by MatthewZhao26** [comment]
@sampan501 not sure why the user guide is still not displaying properly

**Comment by MatthewZhao26** [comment]
> @sampan501 not sure why the user guide is still not displaying properly

@sampan501, I'll try adding a conditional.rst file to the user_guide folder maybe?

**Comment by MatthewZhao26** [comment]
> > @sampan501 not sure why the user guide is still not displaying properly
> 
> @sampan501, I'll try adding a conditional.rst file to the user_guide folder maybe?

@sampan501 should be showing up properly now

**Review by sampan501 — APPROVED** [review]
looks good!

**Pull request merged**